nucypher/blockchain/eth/agents.py

```diff
@@ -688,41 +684,23 @@ def sample(self,
         Only stakers which made a commitment to the current period (in the previous period) are used.
         """
 
-        system_random = random.SystemRandom()
+        # TODO: can be implemented as an iterator if necessary, where the user can
```

Comment:
We need to prioritize bottlenecks of course, but I think this is an excellent idea... not that were ever sampling a ton of addresses at once... or are we?

Reply:
It's not so much about the bottlenecks, but rather about the typical usage scenarios. If the user of `sample()` later discards several of the results for some reason and needs the next best candidates, it is better to make an iterator-like API for it.

Reply:
IIRC, We're sample a buffered amount of nodes at the moment for higher-level fault handling so this sounds like a good iteration.

This ends up being as a finite generator producing ordered sample results.?

Reply:
We might as well include it in this PR.

@KPrasch , where does the list of stakers get filtered? Is it just `BlockchainPolicy.__find_ursulas()`?